src/main/java/org/spongepowered/common/registry/RefreshableRegistry.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.registry;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.google.common.collect.Maps;
+import com.mojang.serialization.Lifecycle;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.spongepowered.common.SpongeCommon;
+import org.spongepowered.common.accessor.resources.ResourceKeyAccessor;
+import org.spongepowered.common.bridge.core.RegistryBridge;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+
+/**
+ * A registry that can be refreshed at runtime,
+ * reloading its contents. Does not support integer
+ * ids as those would be unstable.
+ * @param <T>
+ */
+public class RefreshableRegistry<T> extends Registry<T> {
```

Comment:
I've moved the registries to a new RegistryScope - DATA_PACK. They are now acessable via Sponge.game().dataPackManager().registries()
Currently the actual instance of DataPackManager doesn't change, just the registry inside it, but it functions the same within the API (assuming you don't store DataPackManager).
Let me know what you think of this. I think that scoping them under data pack is the correct way to go.
I also added loadedPacks() to demonstrate that there could easily be more things in here - Reloading it from a plugin, enabling/disabling datapacks?

Reply:
Well, I threw out this as a thought and a question, not necessarily the route we should go down and wanted more of a discussion on it. Specifying a whole new registry holder is a much larger thing than adding a Tag API and it does need some thought.

@Zidane I'd appreciate you weighing in on this as someone who basically built the registry system on top of Vanilla. Do we want to have this notion of a datapack scoped registry? I don't necessarily mind the idea but it is another layer of complexity and as I said above, this does feel a bit like we're forcing the problem into a weird solution (there has to be a reason Mojang didn't go down the registry route here, instead opting for tag collections...) - though it is far better than the "RefreshableRegistry" that was there. 

If that's what we want, we need to think about how we represent this.

Reply:
Just throwing a quick two cents: DataPack scoped registries I think aren’t really feasible from the perspective of when they’re loaded/swapped. In the interest of KISS, a data pack does cause a reload, but it’s still effectively game/server based, right? From the server’s perspective, during a datapack reload, the server’s registries are reloaded for the datapack’s changed/missing options, not something the datapack can explicitly dictate individually right?

Reply:
> In the interest of KISS, a data pack does cause a reload, but it’s still effectively game/server based, right?

Server based.

> From the server’s perspective, during a datapack reload, the server’s registries are reloaded for the datapack’s changed/missing options, not something the datapack can explicitly dictate individually right?

No - updating/reloading registries play no part in the vanilla datapack system, which is kind of my point here. We've been using registries to represent datapack registered values when perhaps we shouldn't be. Advancements, recipes and tags all use their own managers/collections that can be reloaded and so if we were to use the registry system for this, we'd need to replace the registry to rebuild it.

I just don't want to end up forcing this into the registry system when it might not fit right well.

Reply:
So this is the right idea but the wrong approach.

I do think the data pack stuff should have a place in our registry system but everyone needs to remember that data pack content eventually creates registry or registry-like content.

`WorldTypeTemplate` (Data Pack) -> `WorldType` (Registry)

The issue is follow through, making sure we keep doing that. Data Pack types in the API should only be used to build templates and pass to the engine for processing then discorded. Only the built types will matter for registry lookups.

With that out of the way, lets discuss this addition. Vanilla has no registry for tags...no traditional one that is. What they do have is `StaticTagHelper` which...is a stone age registry-like construct. Why Mojang didn't make it a full registry is a question for them...you could likely argue they didn't because going into the registry system now binds you to several contract mechanisms. While I feel that Mojang doesn't make enough registry, API-7 made too many things registry that should not be.

That said, Tags sit on the fence. If you read `StaticTagHelper`, you realize that all `TagTypes` have a `ResourceLocation` as well as all bound `Tag`s have a `ResourceLocation` via the `bind` and `StaticTagHelper.Wrapper` system. In fact, Mojang could have easily did a similar thing you did here...create a `Registry` subclass that can reset itself and register it in the system. Again, question for them why they didn't.

The question we need to answer in Sponge is simply: Do we have this go into the registry framework or simply add some `TagManager` to the API with typical lookups for `Tag` objects that wraps around `StaticTagHelper`. If we are going to stick with the former, your implementation of a `Registry` for them is in-correct, it actually needs to wrap around `StaticTagHelper` to retain mod support.

Reply:
In fact, wrapping around `StaticTagHelper` will basically invalidate any resetting/etc that we are doing. As a bonus, the contract in api-8 for registry means it is legal for the underyling registry to be reset or changed at any time. This is why we do the reference system.

Good idea to keep in mind while discussing this.

Reply:
Wrapped around StaticTagHelper.